protocol/consistencychecks.go

```
@@ -0,0 +1,328 @@
+		return nil, err
+	}
+
+	// 2. verify Auth path
+	proofType, err := verifyAuthPath(uname, key, ap, str)
+	if err != nil {
+		return nil, err
+	}
+
+	// 3. verify (response code, proof type) pair
+	switch {
+	case msg.Error == ErrorNameExisted:
+	case msg.Error == Success && proofType == proofOfAbsence:
+	default:
+		return nil, ErrorMalformedDirectoryMessage
+	}
+
+	// 4. verify returned promise
+	if err := cc.verifyReturnedPromise(requestType,
+		msg.Error, df, uname, key); err != nil {
+		return nil, err
+	}
+
+	if proofType == proofOfAbsence {
+		return str, PassedWithAProofOfAbsence
+	}
+	return str, PassedWithAProofOfInclusion
+}
+
+func (cc *ConsistencyChecks) verifyKeyLookup(requestType int,
+	msg *Response, uname string, key []byte) (*m.SignedTreeRoot, error) {
+
+	df, ok := msg.DirectoryResponse.(*DirectoryProof)
+	if !ok {
+		return nil, ErrorMalformedDirectoryMessage
+	}
+
+	ap := df.AP
+	str := df.STR
+
+	// 1. verify STR
+	if err := cc.verifySTR(cc.SavedSTR, str); err != nil {
+		return nil, err
+	}
+
+	// 2. verify Auth path
+	proofType, err := verifyAuthPath(uname, key, ap, str)
+	if err != nil {
+		return nil, err
+	}
+
+	// 3. verify (response code, proof type) pair
+	switch {
+	case msg.Error == ErrorNameNotFound && proofType == proofOfAbsence:
+	case msg.Error == Success:
+	default:
+		return nil, ErrorMalformedDirectoryMessage
+	}
+
+	// determine which kind of TB's verification we should do
+	// based on the response code and proof type
+	if msg.Error == Success && proofType == proofOfAbsence {
+		// 4.1 verify returned promise
+		if err := cc.verifyReturnedPromise(requestType,
+			msg.Error, df, uname, key); err != nil {
+			return nil, err
+		}
+	} else {
+		// 4.2 verify fulfilled promise
+		if err := cc.verifyFulfilledPromise(uname, str, ap); err != nil {
+			return nil, err
+		}
+	}
+
+	if proofType == proofOfAbsence {
+		return str, PassedWithAProofOfAbsence
+	}
+	return str, PassedWithAProofOfInclusion
+}
```

> (I'll try to come up with an easier to understand alternative after I finish the review)

Yeah, I saw you approved these changes. Is this issue still here?

> I agree, I also pointed this out in #74 (comment)

Do you mean the "fail fast" issue? What do you think about my response?